mods/farming/api.lua

```diff
@@ -7,100 +7,83 @@ local S = farming.get_translator
-	-- check if pointing at a node
-	if not pt then
+function farming.hoe_on_use(itemstack, user, pt, uses)
+	if not pt or pt.type ~= "node" or pt.above.y ~= pt.under.y + 1 then
+		-- Only nodes pointed on the top can be hoed
 		return
 	end
-	if pt.type ~= "node" then
-		return
-	end
-
-	local under = minetest.get_node(pt.under)
-	local p = {x=pt.under.x, y=pt.under.y+1, z=pt.under.z}
-	local above = minetest.get_node(p)
 
-	-- return if any of the nodes is not registered
-	if not minetest.registered_nodes[under.name] then
-		return
-	end
-	if not minetest.registered_nodes[above.name] then
+	if minetest.get_node(pt.above).name ~= "air" then
+		-- The hoe is obstructed from moving if there is no free space
 		return
 	end
 
-	-- check if the node above the pointed thing is air
-	if above.name ~= "air" then
+	local node_under = minetest.get_node(pt.under)
+	local node_under_def = minetest.registered_nodes[node_under.name]
+	if not node_under_def or
+			minetest.get_item_group(node_under.name, "soil") ~= 1 then
+		-- Not a soil node
 		return
 	end
 
-	-- check if pointing at soil
-	if minetest.get_item_group(under.name, "soil") ~= 1 then
+	-- Test if soil properties are defined
+	local soil = node_under_def.soil
+	if not soil or not soil.wet or not soil.dry then
 		return
 	end
 
-	-- check if (wet) soil defined
-	local regN = minetest.registered_nodes
-	if regN[under.name].soil == nil or regN[under.name].soil.wet == nil or regN[under.name].soil.dry == nil then
+	local playername = user:get_player_name()
+	if minetest.is_protected(pt.under, playername) then
+		minetest.record_protection_violation(pt.under, playername)
 		return
 	end
 
-	if minetest.is_protected(pt.under, user:get_player_name()) then
-		minetest.record_protection_violation(pt.under, user:get_player_name())
-		return
-	end
-	if minetest.is_protected(pt.above, user:get_player_name()) then
-		minetest.record_protection_violation(pt.above, user:get_player_name())
-		return
-	end
-
-	-- turn the node into soil and play sound
-	minetest.set_node(pt.under, {name = regN[under.name].soil.dry})
+	-- Put the node which should appear after applying the hoe
+	minetest.set_node(pt.under, {name = node_under_def.soil.dry})
 	minetest.sound_play("default_dig_crumbly", {
 		pos = pt.under,
 		gain = 0.5,
 	}, true)
 
-	if not (creative and creative.is_enabled_for
-			and creative.is_enabled_for(user:get_player_name())) then
-		-- wear tool
-		local wdef = itemstack:get_definition()
-		itemstack:add_wear(65535/(uses-1))
-		-- tool break sound
-		if itemstack:get_count() == 0 and wdef.sound and wdef.sound.breaks then
-			minetest.sound_play(wdef.sound.breaks, {pos = pt.above,
-				gain = 0.5}, true)
-		end
+	if creative and creative.is_enabled_for
+			and creative.is_enabled_for(playername) then
+		return
+	end
+
+	-- wear tool
+	local wdef = itemstack:get_definition()
+	itemstack:add_wear(65535 / (uses - 1))
+	if itemstack:get_count() == 0 and wdef.sound and wdef.sound.breaks then
```

Comment:
(`:is_empty()` would be clearer) and the call to `get_definition` can be moved inside the if.

Reply:
fixed